lint/highlight.py

```diff
@@ -40,304 +37,78 @@ def clear_view(view):
-
-        all.draw(view)
-
-    def line_type(self, line):
-        """Return the primary error type for the given line number."""
-        if not self.all:
-            return None
-
-        line_type = None
-        for highlight in self.all:
-            if line_type == ERROR:
-                continue
-            _line_type = highlight.lines.get(line)
-            if _line_type != WARNING and line_type == WARNING:
-                continue
-            line_type = _line_type
-        return line_type
+def get_line_start(view, line):
+    return view.text_point(line, 0)
 
 
 class Highlight:
     """This class maintains error marks and knows how to draw them."""
 
-    def __init__(self, code=''):
-        self.code = code
-        self.marks = util.get_new_dict()
-        self.mark_style = 'outline'
+    def __init__(self, view):
+        self.view = view
+
+        # Dict[error_type, Dict[style, List[region]]]
+        self.marks = defaultdict(lambda: defaultdict(list))
         self.style_store = HighlightStyleStore()
 
         # Every line that has a mark is kept in this dict, so we know which
         # lines to mark in the gutter.
-        self.lines = util.get_new_dict()
-
-        # These are used when highlighting embedded code
-        # The embedded code is linted as if it begins
-        # at (0, 0), but we need to keep track of where
-        # the actual start is within the source.
-        self.line_offset = 0
-        self.char_offset = 0
-
-        # Linting runs asynchronously on a snapshot of the code.
-        # Marks are added to the code during that asynchronous linting,
-        # and the markup code needs to calculate character positions given
-        # a line + column. By the time marks are added, the actual buffer
-        # may have changed, so we can't reliably use the plugin API to
-        # calculate character positions.
-        # The solution is to calculate and store the character positions for
-        # every line when this object is created, then reference that
-        # when needed.
-        self.newlines = newlines = [0]
-        last = -1
-
-        while True:
-            last = code.find('\n', last + 1)
-
-            if last == -1:
-                break
+        # Dict[error_type, Dict[lineno, style]]
+        self.lines = defaultdict(dict)
 
-            newlines.append(last + 1)
-
-        newlines.append(len(code))
-
-    @staticmethod
-    def strip_quotes(text):
-        """Return text stripped of enclosing single/double quotes."""
-        first = text[0]
-
-        if first in ('\'', '"') and text[-1] == first:
-            text = text[1:-1]
-
-        return text
-
-    def full_line(self, line):
-        """
-        Return the start/end character positions for the given line.
+    def range(self, line, pos, length=-1, near=None, error_type=ERROR, word_re=None, style=None):
+        """Mark a range of text."""
+        raise Exception('`range` has been removed')
```

Comment:
Do we need to keep these around?

Reply:
I guess otherwise any attempts to call it would just fail with an AttributeError.

Reply:
The proper way would probably be to raise a DeprecationWarning, but I'm not sure whether these are ignored by default

Reply:
If it's a deprecation, I should be able to check if any linters still call it and then notify the respective repos. Otherwise just failing with an error is just fine with me, no need to be fancy about this.

Reply:
Ok, not sure if I'm searching correctly, but:

- `highlight.near` has no results
- `highlight.range`:
  - [SublimeLinter-contrib-jsdebugmarkers](https://github.com/LeoLeal/SublimeLinter-jsdebugmarkers/blob/master/linter.py#L99)
  - [SublimeLinter-annotations](https://github.com/SublimeLinter/SublimeLinter-annotations/blob/master/linter.py#L99)
  - SublimeLinter-rst (has been deprecated)
- `self.error`:
  - Again SublimeLinter-contrib-jsdebugmarkers, SublimeLinter-annotations (and SublimeLinter-rst), one line down from `highlight.range`.

So it seems we should update SublimeLinter-annotations and give SublimeLinter-contrib-jsdebugmarkers a notice. @kaste could you perhaps update SublimeLinter-annotations?

Reply:
I couldn't throw a DeprecationWarning bc I already removed it 🍭 

Freudian slip, I knew that annotations used this API - @braver sent me a list of API usage before. But this plugin didn't work on master already. This is just a result of not using the official API but fiddling around. jsdebugmarkers is a fork of annotations. Both reimplement `split_match` within `run` just for fun. I think `run` returns `None`, so SublimeLinter thinks anyway that this lint was clean and does a fast exit. 

Fuck it; I'm really bad at regexes.